File: clients/python/smoke.py
"""Live round trip against a real server, for the sdk-smoke CI job.

    NOVAMEM_SMOKE_URL=… NOVAMEM_SMOKE_TOKEN=… python smoke.py up|down

up:   capture → search finds it → forget deletes it → search no longer finds it.
down: the server has been stopped; search must report UnavailableError, not
      an empty result.
Every failure prints "python <step>: <detail>" and exits 1.
"""

import os
import sys
import uuid
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent / "src"))

import novamem


def fail(step, detail):
    print(f"python {step}: {detail}")
    sys.exit(1)


def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else "up"
    c = novamem.Client(
        os.environ["NOVAMEM_SMOKE_URL"], os.environ["NOVAMEM_SMOKE_TOKEN"]
    )
    if mode == "down":
        try:
            c.search(novamem.SearchRequest(query="anything"))
        except novamem.UnavailableError:
            print("PASS python down")
            return
        except novamem.NovamemError as e:
            fail("down", f"want UnavailableError, got {e!r}")
        fail("down", "search succeeded against a stopped server")

    marker = uuid.uuid4().hex
    try:
        cap = c.capture(
            novamem.CaptureRequest(
                content=f"sdk-smoke python {marker}", namespace="sdk-smoke", force=True
            )
        )
    except novamem.NovamemError as e:
        fail("capture", repr(e))
    if not cap.id:
        fail("capture", f"not saved: {cap!r}")
    hits = c.search(novamem.SearchRequest(query=marker, namespace="sdk-smoke"))
    if cap.id not in [r.id for r in hits.results]:
        fail("search", f"captured {cap.id} not found in {[r.id for r in hits.results]}")
    gone = c.forget(novamem.ForgetRequest(id=cap.id))
    if not gone.deleted:
        fail("forget", repr(gone))
    after = c.search(novamem.SearchRequest(query=marker, namespace="sdk-smoke"))
    if cap.id in [r.id for r in after.results]:
        fail("search-after-forget", f"{cap.id} still returned")
    print("PASS python up")


if __name__ == "__main__":
    main()

File: clients/java/src/test/java/com/azrtydxb/novamem/Smoke.java
// Live round trip against a real server, for the sdk-smoke CI job.
//
//   mvn -q dependency:build-classpath -Dmdep.outputFile=cp.txt
//   NOVAMEM_SMOKE_URL=… NOVAMEM_SMOKE_TOKEN=… \
//     java -cp "$(cat cp.txt):target/classes:target/test-classes" com.azrtydxb.novamem.Smoke
// up|down
//
// up:   capture → search finds it → forget deletes it → search no longer finds it.
// down: the server has been stopped; search must report unavailable, not an
//       empty result. Every failure prints "java <step>: <detail>" and exits 1.

package com.azrtydxb.novamem;

import com.azrtydxb.novamem.types.Types.CaptureRequest;
import com.azrtydxb.novamem.types.Types.CaptureResult;
import com.azrtydxb.novamem.types.Types.ForgetRequest;
import com.azrtydxb.novamem.types.Types.SearchRequest;
import com.azrtydxb.novamem.types.Types.SearchResult;
import java.util.Objects;
import java.util.UUID;
import java.util.function.Supplier;

final class Smoke {
  private Smoke() {}

  private static void fail(String step, String detail) {
    System.out.println("java " + step + ": " + detail);
    System.exit(1);
  }

  private static <T> T step(String name, Supplier<T> f) {
    try {
      return f.get();
    } catch (RuntimeException e) {
      fail(name, e.getMessage());
      throw e;
    }
  }

  public static void main(String[] args) {
    Client c =
        step(
            "connect",
            () ->
                new Client(
                    NovamemConfig.builder()
                        .baseUrl(Objects.toString(System.getenv("NOVAMEM_SMOKE_URL"), ""))
                        .token(Objects.toString(System.getenv("NOVAMEM_SMOKE_TOKEN"), ""))
                        .build()));

    if (args.length > 0 && args[0].equals("down")) {
      try {
        c.search(SearchRequest.builder().query("anything").build());
        fail("down", "search succeeded against a stopped server");
      } catch (NovamemException e) {
        if (!e.isUnavailable()) {
          fail("down", "want unavailable, got " + e.getMessage());
        }
        System.out.println("PASS java down");
        return;
      }
    }

    String marker = UUID.randomUUID().toString().replace("-", "");
    SearchRequest query = SearchRequest.builder().query(marker).namespace("sdk-smoke").build();
    CaptureResult cap =
        step(
            "capture",
            () ->
                c.capture(
                    CaptureRequest.builder()
                        .content("sdk-smoke java " + marker)
                        .namespace("sdk-smoke")
                        .force(true)
                        .build()));
    if (cap.id() == null || cap.id().isEmpty()) {
      fail("capture", "not saved");
    }
    SearchResult hits = step("search", () -> c.search(query));
    if (hits.results().stream().noneMatch(r -> cap.id().equals(r.id()))) {
      fail("search", "captured " + cap.id() + " not found");
    }
    var gone = step("forget", () -> c.forget(ForgetRequest.builder().id(cap.id()).build()));
    if (!Boolean.TRUE.equals(gone.deleted())) {
      fail("forget", "not deleted");
    }
    SearchResult after = step("search-after-forget", () -> c.search(query));
    if (after.results().stream().anyMatch(r -> cap.id().equals(r.id()))) {
      fail("search-after-forget", cap.id() + " still returned");
    }
    System.out.println("PASS java up");
  }
}
